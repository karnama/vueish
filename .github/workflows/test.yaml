name: Test

on:
  pull_request:
    types: [ready_for_review, opened, synchronize, reopened]
    paths:
      - '**/*.ts'
      - '**/*.vue'
      - 'tsconfig*.json'
      - 'jest.config.js'
      - 'package-lock.json'
    branches:
      - main

jobs:
  test:
    if: github.event.pull_request.draft == false
    name: Unit
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
      - uses: actions/cache@v2
        with:
          cache: 'npm'
      - run: npm ci
      - name: Tests
        uses: anuraag016/Jest-Coverage-Diff@master
        with:
          runCommand: "JEST_JUNIT_CLASSNAME='{classname}' JEST_JUNIT_TITLE='{title}' npx jest --coverage --coverageReporters='json-summary' --forceExit --detectOpenHandles --ci --reporters=default --reporters=jest-junit"
          delta: 0.5

      - name: Test Report
        if: ${{ success() || failure() }}
        uses: dorny/test-reporter@v1
        with:
          name: JEST Tests
          path: junit.xml
          reporter: jest-junit
