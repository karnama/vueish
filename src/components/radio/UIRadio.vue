<template>
    <div class="flex ui-radio">
        <label class="relative flex items-center outline-none text-color">
            <input v-bind="$attrs"
                   type="radio"
                   :value="value"
                   tabindex="-1"
                   :disabled="disabled"
                   class="hidden">
            <span class="ui-radio-btn text-brand-500 relative float-left mr-2 h-5 w-5
                         border-2 rounded-full outline-none transition-colors"
                  role="radio" />
            <slot>
                {{ label }}
            </slot>
        </label>
    </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import { label, disabled } from '@/shared-props';

export default defineComponent({
    name: 'UIRadio',

    inheritAttrs: false,

    props: {
        value: {
            type: [String, Boolean, Array, Object, Number],
            required: true
        },

        label,
        disabled
    }
});
</script>

<style lang="scss" scoped>
.ui-radio-btn:after {
    content: '';
    position: absolute;
    top: 3px;
    left: 3px;
    width: 10px;
    height: 10px;
    border-radius: 100%;
    background: currentColor;
    transform: scale(0);
    transition: all .2s ease;
    opacity: .08;
    pointer-events: none;
}

label {
    cursor: pointer;

    &.disabled {
        cursor: not-allowed;

        .ui-radio-btn:after {
            background: theme('colors.gray.500');
        }
    }
}

input:checked + .ui-radio-btn:after {
    transform: scale(1) !important;
    transition: all .2s cubic-bezier(.35, .9, .4, .9);
    opacity: 1;
}

input:checked:enabled + .ui-radio-btn {
    border-color: currentColor;
}

label:hover input:enabled > .ui-radio-btn:after {
    transform: scale(3.6);
}

label:focus > input:enabled + .ui-radio-btn {
    border-color: currentColor;
}
</style>
