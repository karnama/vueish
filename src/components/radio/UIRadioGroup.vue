<template>
    <div class="ui-radio-group">
        <span class="label text-sm text-gray-500">
            <slot name="label">
                {{ label }}
            </slot>
        </span>

        <div ref="slot"
             class="slot flex pt-2"
             role="radiogroup"
             :class="horizontal ? 'space-x-6': 'flex-col space-y-4'">
            <slot />
        </div>
    </div>
</template>

<script lang="ts">
import { defineComponent, ref, onMounted, onBeforeUpdate, watch } from 'vue';
import { label, disabled, name } from '@composables/input';

export default defineComponent({
    name: 'UIRadioGroup',

    props: {
        modelValue: {
            type: [String, Boolean, Number, Object, Array]
        },

        label,
        disabled,
        name,
        horizontal: Boolean,
        required: Boolean
    },

    emits: ['update:modelValue'],

    setup(props, ctx) {
        const slot = ref<HTMLDivElement>();
        const watchStopHandlers: ReturnType<typeof watch>[] = [];
        const updateInputs = () => {
            watchStopHandlers.forEach(stop => stop());
            watchStopHandlers.length = 0;

            // Search for radio inputs in the slot
            const inputs = slot.value?.querySelectorAll('input[type="radio"]') as NodeListOf<HTMLInputElement>;

            // Validate they exist
            if (!inputs || inputs.length < 2) {
                throw Error('UIRadioGroup requires at least 2 UIRadio components present.');
            }

            const setChecked = (input: HTMLInputElement) => input.checked = props.modelValue === input.value;
            const setName = (input: HTMLInputElement) => {
                input.name = props.name!;
            };
            const setDisabled = (input: HTMLInputElement) => {
                const disabled = input.hasAttribute('disabled') ? input.disabled : props.disabled;
                const label = input.closest('label')!;
                input.disabled = disabled;

                if (disabled) {
                    label.classList.add('disabled');
                } else {
                    label.classList.remove('disabled');
                }

                label.tabIndex = input.disabled ? -1 : 0;
            };

            watchStopHandlers.push(
                // Update the inputs on props change
                watch(() => props.name, () => inputs.forEach(setName), { immediate: true }),
                watch(() => props.modelValue, () => inputs.forEach(setChecked), { immediate: true }),
                watch(() => props.disabled, () => inputs.forEach(setDisabled), { immediate: true }),
                watch(() => props.required, value => inputs[0].required = value, { immediate: true })
            );

            // Set event listeners
            inputs.forEach(input => {
                input.onclick = () => ctx.emit('update:modelValue', input.value);
                input.closest('label')!.onkeydown = (e: KeyboardEvent) =>
                    e.key !== ' ' || ctx.emit('update:modelValue', input.value);
            });
        };

        onMounted(updateInputs);
        onBeforeUpdate(updateInputs);

        return { slot };
    }
});
</script>
