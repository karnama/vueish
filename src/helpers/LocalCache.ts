/**
 * localStorage wrapper utility to manage values safely.
 */
export default class LocalCache {
    /**
     * The string value to be added to every keys to avoid clash.
     *
     * @private
     */
    private prefix = '';

    /**
     * The class constructor.
     *
     * @param namespace
     *
     * @return {LocalCache}
     */
    public constructor(namespace = '') {
        this.namespace = namespace;

        return this;
    }

    /**
     * The get method.
     *
     * @param key
     * @param defaultValue
     *
     * @return {*}
     */
    public get<T extends any>(key: string, defaultValue?: T): T | undefined {
        try {
            let value = JSON.parse(localStorage.getItem(this.prefix + key) as string);

            // returns null if not defined
            if (value === null || value === undefined) {
                value = defaultValue;
            }

            return value;
        } catch (e) {
            // eslint-disable-next-line no-console
            console.error(
                '"' + key + '" cannot be parsed to a Javascript type: \n'
                + String(localStorage.getItem(this.prefix + key))
            );
            this.remove(key);

            return;
        }
    }

    /**
     * Set the value in local storage
     *
     * @param {String} key
     * @param {*} value
     *
     * @return {LocalCache}
     */
    public set(key: string, value: any): this {
        localStorage.setItem(this.prefix + key, JSON.stringify(value));

        return this;
    }


    /**
     * Merge the value with existing value.
     *
     * @param {String} key
     * @param {*} value
     *
     * @return {LocalCache}
     */
    public add(key: string, value: any): this {
        let storedValue = this.get(key);

        if (!storedValue) {
            this.set(key, value);

            return this;
        }

        if (Array.isArray(storedValue)) {
            if (!Array.isArray(value)) {
                value = [value];
            }

            storedValue.push(...value);
        } else if (storedValue === Object(storedValue) && value === Object(value)) {
            storedValue = Object.assign(storedValue, value);
        } else {
            // default to Javascript's type coercion
            storedValue += value;
        }

        this.set(key, storedValue);
        return this;
    }

    /**
     * Remove the value from localStorage.
     *
     * @param {String} key
     *
     * @return {LocalCache}
     */
    public remove(key: string): this {
        localStorage.removeItem(this.prefix + key);
        return this;
    }

    /**
     * Determines whether the key exists or not.
     *
     * @param {String} key
     *
     * @return {Boolean}
     */
    public has(key: string): boolean {
        return !!this.get(key);
    }

    /**
     * The fluent setter for the namespace
     * @param namespace
     */
    public set namespace(namespace: string) {
        if (typeof namespace !== 'string') {
            throw new Error('Invalid type for namespace. Expected string got ' + typeof namespace);
        }

        this.prefix = namespace + '_';
    }
}
