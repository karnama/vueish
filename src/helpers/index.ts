import { getCurrentInstance } from 'vue';
import type { Icon, Settings } from '@/types';

export function getLibrarySettings(): Settings {
    return getCurrentInstance()?.appContext.config.globalProperties.Vueish;
}

export function getIcon(icon: Icon): string {
    return getLibrarySettings().icons[icon];
}
