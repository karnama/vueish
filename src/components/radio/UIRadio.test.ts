import { mount } from '@vue/test-utils';
import type { DOMWrapper } from '@vue/test-utils';
import UIRadio from './UIRadio.vue';
import UIRadioGroup from './UIRadioGroup.vue';
import { disableConsoleWarn, enableConsoleWarn } from 'helpers/test';

const selectorMap = {
    radioGroup: {
        label: 'span.font-medium.text-color.flex.items-center'
    }
} as const;

describe('UIRadio', () => {
    it('should bind attributes to the input', () => {
        const foo = 'bar';

        const wrapper = mount(UIRadio, {
            props: {
                value: 1
            },
            attrs: {
                foo
            }
        });

        expect(wrapper.get('input').attributes().foo).toBe(foo);
    });

    it('should display the label when passed as a prop', () => {
        const label = 'text';

        const wrapper = mount(UIRadio, {
            props: {
                value: 1,
                label
            }
        });

        expect(wrapper.get('label').text()).toBe(label);
    });

    it('should display the label when passed as a slot', () => {
        const label = 'text';

        const wrapper = mount(UIRadio, {
            props: {
                value: 1
            },
            slots: {
                default: label
            }
        });

        expect(wrapper.get('label').text()).toBe(label);
    });
});

describe('UIRadioGroup', () => {
    it('should display correctly', () => {
        const wrapper = mount({
            template: `
                <UIRadioGroup label="my-label" name="name">
                <UIRadio value="foo"/>
                <UIRadio value="bar"/>
                </UIRadioGroup>`,
            components: { UIRadioGroup, UIRadio }
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it('should display the label when passed as a prop', () => {
        const label = 'text';

        const wrapper = mount({
            template: `
                <UIRadioGroup label="${label}" name="name">
                  <UIRadio value="foo" />
                  <UIRadio value="bar" />
                </UIRadioGroup>`,
            components: { UIRadioGroup, UIRadio }
        });

        expect(wrapper.get(selectorMap.radioGroup.label).text()).toBe(label);
    });

    it('should display the label when passed as a slot', () => {
        const label = 'text';

        const wrapper = mount({
            template: `
                <UIRadioGroup name="name">
                <template #label>${label}</template>
                <UIRadio value="foo"/>
                <UIRadio value="bar"/>
                </UIRadioGroup>`,
            components: { UIRadioGroup, UIRadio }
        });

        expect(wrapper.get(selectorMap.radioGroup.label).text()).toBe(label);
    });

    it('should bind the correct classes when in default orientation', () => {
        const wrapper = mount({
            template: `
                <UIRadioGroup name="name">
                <UIRadio value="foo"/>
                <UIRadio value="bar"/>
                </UIRadioGroup>`,
            components: { UIRadioGroup, UIRadio }
        });

        expect(wrapper.get('.ui-radio-group .slot').classes()).toContain('flex-col');
    });

    it('should bind the correct classes when set to horizontal', () => {
        const wrapper = mount({
            template: `
                <UIRadioGroup horizontal name="name">
                <UIRadio value="foo"/>
                <UIRadio value="bar"/>
                </UIRadioGroup>`,
            components: { UIRadioGroup, UIRadio }
        });


        expect(wrapper.get('.ui-radio-group .slot').classes()).not.toContain('flex-col');
    });

    it('should bind the name value to the radio inputs', () => {
        const wrapper = mount({
            template: `
                <UIRadioGroup name="input">
                <UIRadio value="foo"/>
                <UIRadio value="bar"/>
                </UIRadioGroup>`,
            components: { UIRadioGroup, UIRadio }
        });

        const inputs = wrapper.findAll('input[type="radio"]');
        inputs.forEach(input => expect(input.attributes().name).toBe('input'));
    });

    it('should bind the disabled value to the radio inputs', async () => {
        const wrapper = mount({
            template: `
                <UIRadioGroup name="name">
                <UIRadio value="foo"/>
                <UIRadio value="bar"/>
                </UIRadioGroup>`,
            components: { UIRadioGroup, UIRadio }
        });

        const inputs = wrapper.findAll('input[type="radio"]');

        inputs.forEach(input => expect(input.attributes().disabled).toBeUndefined());
        await wrapper.setProps({ disabled: true });
        inputs.forEach(input => expect(input.attributes().disabled).toBeDefined());
    });

    it('should bind the required value to the radio inputs', async () => {
        const wrapper = mount({
            template: `
                <UIRadioGroup name="name">
                <UIRadio value="foo"/>
                <UIRadio value="bar"/>
                </UIRadioGroup>`,
            components: { UIRadioGroup, UIRadio }
        });

        const inputs = wrapper.findAll('input[type="radio"]');

        expect(inputs[0].attributes().required).toBeUndefined();
        await wrapper.setProps({ required: true });
        expect(inputs[0].attributes().required).toBeDefined();
    });

    it('should bind the modelValue to the radio inputs', async () => {
        const wrapper = mount({
            template: `
                    <UIRadioGroup>
                    <UIRadio value="foo"/>
                    <UIRadio value="bar"/>
                    </UIRadioGroup>`,
            components: { UIRadioGroup, UIRadio }
        },
        {
            props: {
                name: 'name',
                modelValue: 'foo'
            }
        });

        const inputs: DOMWrapper<HTMLInputElement>[] = wrapper.findAll('input[type="radio"]');

        expect(inputs[0].element.checked).toBe(true);
        expect(inputs[1].element.checked).toBe(false);
        await wrapper.setProps({ modelValue: 'bar' });
        expect(inputs[0].element.checked).toBe(false);
        expect(inputs[1].element.checked).toBe(true);
    });

    it('should update the modelValue when a child is clicked', async () => {
        const wrapper = mount({
            template: `
                    <UIRadioGroup>
                    <UIRadio value="foo"/>
                    <UIRadio value="bar"/>
                    </UIRadioGroup>`,
            components: { UIRadioGroup, UIRadio }
        },
        {
            props: {
                name: 'name',
                modelValue: 'foo'
            }
        });

        const radioGroup = wrapper.findComponent(UIRadioGroup);
        const input = wrapper.find('input[type="radio"]');

        await input.trigger('click');
        expect(radioGroup.emitted('update:modelValue')).toBeDefined();
        expect(radioGroup.emitted('update:modelValue')![0]).toStrictEqual(['foo']);
    });

    it('should throw an error if there are missing UIRadio components', () => {
        disableConsoleWarn();
        expect(() => mount(UIRadioGroup))
            .toThrow(new RangeError('UIRadioGroup requires at least 2 UIRadio components in the default slot.'));

        enableConsoleWarn();
    });
});
