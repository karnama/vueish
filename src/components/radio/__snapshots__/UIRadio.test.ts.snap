// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIRadioGroup should display correctly 1`] = `
<div
  class="ui-radio-group"
>
  <transition-stub>
    
    <span
      class="font-medium text-color inline-flex items-center mb-1.5"
    >
      
      my-label
      
    </span>
    
  </transition-stub>
  <div
    class="slot flex w-full flex-col space-y-4"
    role="radiogroup"
  >
    
    <div
      class="flex ui-radio"
    >
      <label
        class="relative flex items-center outline-none text-color"
        tabindex="0"
      >
        <input
          class="hidden"
          name="name"
          tabindex="-1"
          type="radio"
          value="foo"
        />
        <span
          class="ui-radio-btn text-brand-500 relative float-left mr-2 h-5 w-5 border-2 rounded-full outline-none transition-colors"
          role="radio"
        />
        
        
        
      </label>
    </div>
    <div
      class="flex ui-radio"
    >
      <label
        class="relative flex items-center outline-none text-color"
        tabindex="0"
      >
        <input
          class="hidden"
          name="name"
          tabindex="-1"
          type="radio"
          value="bar"
        />
        <span
          class="ui-radio-btn text-brand-500 relative float-left mr-2 h-5 w-5 border-2 rounded-full outline-none transition-colors"
          role="radio"
        />
        
        
        
      </label>
    </div>
    
  </div>
  <transition-stub>
    
    
  </transition-stub>
</div>
`;
